Better File manager menu, adding multi upload file properties and menu customization from addons

goutnet opened the pull request:
This branch perform quite a lot of improvements on the file manager, and add the ability to call any of the bulk  actions (now converted to a drop down menu) onto the freshly uploaded files.

Most of the conversation is here #537.

@aembler depending on your feedback, this might be ready to go.

goutnet commented:
Ok depending on your feedback, this one is now good to go.

What's in:

- Added a FileManagerMenu helper to allow centralization and customization of the file manager menu
- Each action is now customizable, and can have a controller to run in the background (the very same way it is done for the dashboard menu).
- Refactored the bulk menu to use the few FileManagerMenu helper
- Refactored the context menu to use that new helper
- moved the bulk menu to a Bootstrap dropdown with a target button (see next item)
- recently uploaded files now share the bulk menu, a button had been added to let you choose between selected or uploaded
- added icons support for each menu items (both bulk menu and context menu)
- Added a 'dangerous' marking for Delete (and available for any new action in the menu)
- download tool is now a route
- delete tool is deleted (since there is already a route doing that)
- harmonized all the parameters of the file menu actions responder (file id(s) are thrown in the fID _REQUEST variable)

Both menus now look like this

![both-menus](https://cloud.githubusercontent.com/assets/3543067/5624105/ff91839c-955f-11e4-9dbf-5b4c8fe62eb3.png)

goutnet commented:
I just found that I broke the file manager when called from a file selector.

I am fixing it and refactoring a bit more the js to prevent those issue, sorry for the inconvenience.

goutnet commented:
ok this is fixed now. The file manager template is outputted with the selector, although I shoud probably move it to an asset, this works for now (so you can at least review it).

goutnet commented:
Thanks for the reply, some comments on it:

> I'm not 100% sold on the menu refactoring. I do actually think there's potential for refactoring our file and page menus – perhaps with some ability to modify the menus from a package or from some custom code, which it looks like you have, and I think is really cool. But I don't really like how the menu is an element that now has to be included along with concreteFileSelector and the search.

Actually it does not have to, it is included directly in it :)

Now, I think I could make the template into an asset that is automatically pulled when you get the js with it (but that is the point of #1773).

>  Now, if I run the concreteFileSelector JavaScript it's not going to work without me remembering to use the file_context_menu element. I think it was more elegant to have it be part of that JavaScript.

Well, or having it as an asset automatically pulled for you… which is what I suggest in #1773 

> Also, I'm not sure that just opening the bulk menu when uploading files makes total sense. I mean, it does to me, and I like the elegance of the solution – but at the same time there are a lot of options in that menu that don't make sense in the context of a recently uploaded file. I'm not going to need to rescan it or download it or delete it, for example.

Ok maybe rescan would be not really needed, but download to check your upload, or delete for undoing your upload is definitely a must. I mean, how many time did you drag the wrong file and cursed you for doing it? Now just hit the delete button and your done (if you don't do it for you, do it for the clumsy users like me :D).

All in all, if I add 'template' layer of assets, and make the menu into an asset, this is all solved, am I right?

Additionally do you want me to get rid of the tools and continue moving them to routes? (I will do that after we settle the template/refactoring thing) ?

goutnet commented:
Oh, and removing the icons, well very very very simple, if you want they are as good as gone.

That said, I am thinking about the support here, if you are supporting a platform that is not in your language, say… Chinese … having icons to help you actually makes a point (don't understand the text, no brainer, the icon is still here).

So what do you want me do to with the icons? 

(I can also just remove the icons in the default menu, but keep the possibility of adding icons for menu items added from packages or application for instance ?)

goutnet commented:
> The selected/uploaded menu control is interesting. I do like how the new solution doesn't really put a big dialog window in front of the user after they upload any file – but at the same time I kind of miss the old dialog window. 

Well, 5.7 is lighter in its UI, and I think this modification goes the right way.

> For one thing, the reason we added it in the first place was to really give users the idea that their files could be categorized and modified after uploading. I think we kind of failed at that (since sets were buried on the third tab) but I think we could do better. 

Definitely, that was not working, I mean, how many times I uploaded files, closed the dialog by accident not realizing that it would be a hell of a pain to modify properties/filset *after*

> Additionally, at one point we had wanted to add some functionality to FORCE users to categorize files before they would actually be 100% added into the file manager. I'm not saying that we're going to do that but the dialog window approach to showing the meta data and the file sets would give us a spot to really force that.

I don't think this is a good idea, I mean, a file manager is supposed to help user manage their files and provide ways for it. Not FORCE any workflow.

> Additionally, at one point uploaded files would get a "Choose" button if the file manager were in select mode. I definitely miss that – since users have to click "Choose" on the file they select to get it into the image block, for example. 

Well, here again, the Choose menu seriously was a bad ui IMO, what you expect is having to choose in one click, not 2. I really enjoy the new way having the file menu available directly in the image without having to fire the file manager to access properties etc… (also having  a responsive way to view images - #1774 would help smoothing the process). Do you want me to add it?

> Also, I'm not sure that just opening the bulk menu when uploading files makes total sense. I mean, it does to me, and I like the elegance of the solution – but at the same time there are a lot of options in that menu that don't make sense in the context of a recently uploaded file. I'm not going to need to rescan it or download it or delete it, for example.

Also one more thing, having this menu is much more scalable, every new operation on multiple files can be accessible to uploaded files at no cost. And the uploaded file "remembers" (sort of) the ids so you can work on your selected files, then come back and check on your uploaded files again. That means it is a lot more error tolerant for the user, they don't have to get it right the first time, they can always come back in a second time (as long as they don't leave the page that is).

ahukkanen commented:
Not straightly related to this one but I thought this might be the right place to raise this up:
In previous concrete5 versions, it was possible to select multiple files with the file manager (through the checkboxes and then selecting "Choose" from the dropdown). In 5.7 this is no longer possible because the first click already selects the file.

I'm not sure if this is a bug or intended behavior. I might also be better in its own issue thread here but since this one does some refactoring, it might be also relevant to address this issue.

This was a useful feature for us in multiple places.

By the way, really nice and clean solution to managing the uploaded files!

ahukkanen commented:
And above I mean when the file selector is used in blocks or single pages for instance. E.g. you could have list of whiles where you want to select multiple files to. Now you need to re-open the file manager every time and add the files one by one.

goutnet commented:
Hum, given the way the javascript is laid out, it seems to be intended to me, but we can definitely bring that behavior back if needed.

ahukkanen commented:
I definitely think it would be very useful in many cases but let's hear first from @aembler.

aembler commented:
I actually liked the icons back when we had them in 5.6 and earlier but during the 5.7 process we were advised to remove them, so now we should probably stay consistent.

goutnet commented:
Well, I will remove the icons, but let the possibility to have them

goutnet commented:
I removed the icons, but I let the possibility to have them.

Later when (if) this gets merged, I will create an addon that sets the icons for free on the marketplace to let users have icons if they want it. Are you ok with this?

Also, I'd like you to merge #1785 so I can put a template that is pulled together with the file manager selector or manager and remove the need to use the element from js. (this would solve the issue you mentionned).

ahukkanen commented:
As of my point regarding selecting multiple files through the File Manager, I created another issue thread: #1799.

ahukkanen commented:
@goutnet See #1799, @aembler agreed that the bulk selection would be good to bring back.

I could also take this under my belt but since this touches a lot in the File Manager, maybe it's better to join into this one. Or what do you thing?

goutnet commented:
Yeah I can add it here, no problem. 

Now, since I don't really have a direction yet from @aembler, I am not sure exactly where to go from this point.

My understanding is that andrew is not really against this neither for this pull request, and he is seriously busy into things on portland's lab atm.

so I can add more  to this (and I am all for it), but I am not sure it would help andrew to review it more easilly.

@aembler, I do understand that your time is precious (mine too to be honest…), but could you spare 5min to tell me how I could help you ease the process of reviewing this one? Do you want me to write you a architecture doc, something like that?

aembler commented:
@goutnet That's true, but it kind of underscores a problem with this pull request. Had it been just been something that addressed the original issue, which was always well defined and simply presented, it would have been fine and honestly probably would have made it in for 5.7.3.1. It's not simple, though; it's a complete refactoring and reworking of multiple menu systems, introduces new major concepts and a completely new interface paradigm that we don't use anywhere else, and even then doesn't reintroduce all the functionality that we were trying to bring back from 5.6. It's for these reasons that we're on the fence about including it, and of course anything else that builds off of it.

goutnet commented:
@ambler: It does, that was the point actually…

I did have that talk with Franz and this is why I am saying that, I personally think that this pull request will make you coffee and bring you the soda you did not think of, although I promise to first talk of my next ideas before I go again for a big automatic soda machine (Franz should understand that last one :) )

Well, jokes aside, and because fences are to be lifted to the right incoming, I understand you are busy, I understand that this pull request fixes stuffs, but also add new stuffs you did not expected.

You say I am not bringing back all the features you wanted from 5.6. Can you make a simple bullet list of what you would like me to bring more with this pull request? (don't be afraid to be very specific here)? I am opened to any suggestion here.

goutnet commented:
also regarding the js template, I will make it into an ajax request directly pulled by the manager if no menu is around (cleaner).

goutnet commented:
I moved the menu to a route, added a view for it (using the same element to avoid copy/paste).

Now the js will pull this menu from the route if no menu is found in the page, so no need to have the element included anymore. This way, even if you call the javascript directly, then the menu will work seemlessly, plus it lets you change the menu completely if you feel like it by changing the template variable.

At this point, I am really needing some input from you guys (c5 team), what is left on my todolist is this:

- Bring back the possibility to select multiple files from the selector
- Add a target for "duplicated" files? (issue still unanswered)
- Add a responsive toggle to image views (to have them resize their view window), activated by default

Not sure about what other features you need me to bring back…

goutnet commented:
As requested in #1774 I moved the view into a route (removing the tool) then made the view content responsive by default:

On the way, since it made no sense to add a controller just for this, I added a new method: Conrtoller::setViewPath($path) to allow a Controller to set its own view after object construct.


![resized-view](https://cloud.githubusercontent.com/assets/3543067/5771769/27163a06-9d45-11e4-8527-7230a07585ac.png)
![non-resized-view](https://cloud.githubusercontent.com/assets/3543067/5771768/2710e36c-9d45-11e4-972d-ee13b3ba546e.png)

EC-Joe commented:
> You say I am not bringing back all the features you wanted from 5.6. Can you make a simple bullet list of what you would like me to bring more with this pull request? (don't be afraid to be very specific here)? I am opened to any suggestion here.

+1 to the list. It's hard to hit a target with no definition of the task at hand. Also, with as big as this is getting, perhaps it might be best to kick this over into a feature branch instead of trying to get it straight into develop?

goutnet commented:
what is the difference between a feature branch and a branch on sb else fork ? (which is really only a pull request there)… not sure I see any value to it :/

goutnet commented:
@aembler After a second thought, I prefer to stop here. Bringing the other features back (as well as removing legacy tools) will hapen in another pull request that should build on top of that one.

To avoid merging conflict as well as to add complexity to this pull request, I will patiently (or not?) wait for you to get into this one and hopefully merge it. (I had to fix a merge conflict because of the waiting time here, and because you did not pay attention to the comments I placed on that pull request as well, so this is the last straw for me here).

I think it adds quite a bit of features here, and brings up a pretty nice way of working with the file menu (all of that in the same way c5 general architecture is meant). As far as it goes, I addressed all your concerns, improved the overall system, refactored what's needed to be etc…

I also believe I have put enough efforts in this to be considered worthy of a review on your side, and that wasting more of my time on something without even having a goal properly defined is simply not imaginable.

Just a last word here, You guys explained you are busy, alright,  I am quite too, and probably a *lot* more than you actually are (I mean, seriously, unless you guys hit the 100 work hours/week you just not even trying to match my time here). I work during nites, weekend and holidays for you, cutting down my sleeping time quite a bit to bring you some free work, Think about it a bit would you?. 

I asked for a bit of your time, or at least clear directions, all I got was nothing or some pretty sparse comments, only leaving me the option of either going forward into "imagining" what you might want to see, or stopping. I took the first option so far, but since nothing happen, I am forced into the second.

This is as far as I go… I hope to have some more refreshing news in the future.

aembler commented:
That's too bad. I thought the original issue was clearly written, and the follow–up responses, while perhaps trying too hard to be complimentary, expressed the reservations we had about the wide-reaching aspects of all this new code. There are interesting concepts here that, had they been first discussed and then implemented in their own, self-contained pull requests, likely would have made it in. Unfortunately, this pull request is both too complicated and doesn't do enough to actually address the original issue to be considered as it currently stands. Hopefully we can address the issue again soon.

goutnet commented:
Well, I am sad to read this answer (especially the part that says it does not address the original issue that was actually totally solved by this pull request). I will then open tickets for each thing I did already implemented in that pull request and try to break them down in multiple pull request as you wish it… (feel like wasting time in discussions but well…).

At least I hope to be able to merge those in an order that will not require me to create code that would be trashed by a next to come request.

goutnet commented:
mainly my issue is that all of the opened issues (including the nitial one) would have to stack on each other, so I would have to wait for each one to be merged before I could start to work on the next).

Anyway, with your consent I would start on newly opened #1848, which would be the base for the rest (and make the rework less painful - but still painful anyway)

aembler commented:
Yes, let's start with that one.

On Wed, Jan 21, 2015 at 8:30 AM, Florian Delizy <notifications@github.com>
wrote:

> mainly my issue is that all of the opened issues (including the nitial
> one) would have to stack on each other, so I would have to wait for each
> one to be merged before I could start to work on the next).
>
> Anyway, with your consent I would start on newly opened #1848
> <https://github.com/concrete5/concrete5-5.7.0/issues/1848>, which would
> be the base for the rest (and make the rework less painful - but still
> painful anyway)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/concrete5/concrete5-5.7.0/pull/1762#issuecomment-70869594>
> .
>
